### PR TITLE
feat(plugin): Allow pre-release versions of the plugin

### PR DESCRIPTION
**Pull request by AndyTWF**
Allows us to release more incrementally, large features.

### CODE REVIEW

File: app/Http/Controllers/VersionController.php [hunk 1]
```diff
@@ -67,7 +82,7 @@ public function createNewPluginVersion(Request $request): JsonResponse
 
         try {
             $this->versionService->publishNewVersionFromGithub($request->json('release.tag_name'));
-        } catch (VersionAlreadyExistsException $alreadyExistsException) {
+        } catch (VersionAlreadyExistsException|ReleaseChannelNotFoundException) {
```
Comment (AxonC) [inline app/Http/Controllers/VersionController.php:85]: This 8.1? Like it if so.

File: app/Services/VersionService.php [hunk 2]
```diff
@@ -42,36 +47,16 @@ public function getVersion(string $versionString) : Version
...
-
-        return $version->wasRecentlyCreated ? true : false;
-    }
-
     /**
      * Toggles whether or not a version is allowed to be used in production.
      *
-     * @throws VersionNotFoundException
      * @return void
+     * @throws VersionNotFoundException
```
Comment (AxonC) [inline app/Services/VersionService.php:59]: Docstring doesn't document all thrown exceptions.

File: app/Services/VersionService.php [hunk 3]
```diff
@@ -90,15 +75,68 @@ public function publishNewVersionFromGithub(string $tag)
...
+            throw new ReleaseChannelNotFoundException();
+        }
+
+        $releaseChannel = PluginReleaseChannel::where('name', VersionParser::parseStability($normalisedVersion))
+            ->first();
+
+        if (!$releaseChannel) {
+            Log::error(sprintf('Invalid release channel %s', $tag));
+            throw new ReleaseChannelNotFoundException();
+        }
```
Comment (AxonC) [inline app/Services/VersionService.php:91]: Could be simplified by using `firstOrFail` and catching the `ModelNotFoundException`?

File: app/Services/VersionService.php [hunk 4]
```diff
@@ -90,15 +75,68 @@ public function publishNewVersionFromGithub(string $tag)
...
+
+    /**
+     * The "latest version" on any release channel is the most recent version (in semver terms)
+     * on any channel that is, or more stable than, the requested channel.
+     */
+    public function getLatestVersionForReleaseChannel(string $channel): Version
+    {
+        $relevantVersions = $this->getRelevantVersions(PluginReleaseChannel::where('name', $channel)->firstOrFail());
+        if ($relevantVersions->isEmpty()) {
+            throw new VersionNotFoundException();
```
Comment (AxonC) [inline app/Services/VersionService.php:123]: Missing @throws and @returns on docstring (only pointing out for consistency in file).

File: app/Services/VersionService.php [hunk 5]
```diff
@@ -90,15 +75,68 @@ public function publishNewVersionFromGithub(string $tag)
...
-        Version::where('id', '<>', $newVersion->id)->delete();
+        Version::where('id', '<>', $newVersion->id)
+            ->get()
+            ->reject(function (Version $version) use ($newVersion) {
+                return Comparator::greaterThan(
+                    $version->version,
+                    $newVersion->version
+                );
+            })
+            ->each(function (Version $version) {
```
Comment (AxonC) [inline app/Services/VersionService.php:110]: Arrow function in lieu of new php version for this?

File: app/Services/VersionService.php [hunk 6]
```diff
@@ -90,15 +75,68 @@ public function publishNewVersionFromGithub(string $tag)
...
+            ) ? $version : $selectedVersion;
+        });
+    }
+
+    private function getRelevantVersions(PluginReleaseChannel $requestedChannel): Collection
+    {
+        return Version::whereHas('pluginReleaseChannel', function (Builder $releaseChannel) use ($requestedChannel) {
+            return $releaseChannel->where('relative_stability', '<=', $requestedChannel->relative_stability);
+        })
+            ->get();
```
Comment (AxonC) [inline app/Services/VersionService.php:139]: Seems a little bit odd on its own line :D

File: app/Http/Controllers/VersionController.php (hunk 1 above)
Comment (AndyTWF) [reply app/Http/Controllers/VersionController.php:85]: That it be!

File: app/Services/VersionService.php (hunk 2 above)
Comment (AndyTWF) [reply app/Services/VersionService.php:59]: I think it only throws this one?

File: app/Services/VersionService.php (hunk 3 above)
Comment (AndyTWF) [reply app/Services/VersionService.php:91]: Thought about this - I'm thinking to keep as we need the same exception class further down. Plus saves us having to catch another exception in the controller and Github gets grumpy if you respond with 404 to its web hooks :D

File: app/Services/VersionService.php (hunk 4 above)
Comment (AndyTWF) [reply app/Services/VersionService.php:123]: Done!

File: app/Services/VersionService.php (hunk 5 above)
Comment (AndyTWF) [reply app/Services/VersionService.php:110]: Oooh, go on then

File: app/Services/VersionService.php (hunk 6 above)
Comment (AndyTWF) [reply app/Services/VersionService.php:139]: Made it less lonely

File: app/Services/VersionService.php (hunk 2 above)
Comment (AxonC) [reply app/Services/VersionService.php:58]: ReleaseChannelNotFoundException is also thrown unless I'm going mad

File: app/Services/VersionService.php (hunk 3 above)
Comment (AxonC) [reply app/Services/VersionService.php:94]: Sound reasoning - cheers

File: app/Services/VersionService.php (hunk 5 above)
Comment (AxonC) [reply app/Services/VersionService.php:111]: You can probably do the each callback too, for extra goodness

File: app/Services/VersionService.php (hunk 2 above)
Comment (AndyTWF) [reply app/Services/VersionService.php:58]: Don't think it does!

File: app/Services/VersionService.php (hunk 5 above)
Comment (AndyTWF) [reply app/Services/VersionService.php:111]: Not sure if I can - fn functions returns the result of expr, which could short-circuit the each if they're falsey :)

**Review by AxonC — APPROVED** [review]

**Pull request merged**